cmake/modules/python.cmake

```diff
@@ -13,35 +13,51 @@ endif()
 
 set(PYTHON_MINIMUM_REQUIRED 3.8)
 
-# We are using foreach here, instead of find_program(PYTHON_EXECUTABLE_SYSTEM_DEFAULT "python" "python3")
-# cause just using find_program directly could result in a python2.7 as python, and not finding a valid python3.
-foreach(PYTHON_PREFER ${PYTHON_PREFER} ${WEST_PYTHON} "python" "python3")
-  find_program(PYTHON_PREFER_EXECUTABLE ${PYTHON_PREFER})
-  if(PYTHON_PREFER_EXECUTABLE)
-      execute_process (COMMAND "${PYTHON_PREFER_EXECUTABLE}" -c
-                               "import sys; sys.stdout.write('.'.join([str(x) for x in sys.version_info[:2]]))"
-                       RESULT_VARIABLE result
-                       OUTPUT_VARIABLE version
-                       ERROR_QUIET
-                       OUTPUT_STRIP_TRAILING_WHITESPACE)
+function(python_get_version interpreter)
+  execute_process(COMMAND "${interpreter}" -c
+    "import sys; sys.stdout.write('.'.join([str(x) for x in sys.version_info[:2]]))"
+    RESULT_VARIABLE result
+    OUTPUT_VARIABLE version
+    ERROR_QUIET
+    OUTPUT_STRIP_TRAILING_WHITESPACE)
+  set(PYTHON_VERSION "${version}" PARENT_SCOPE)
+endfunction()
 
-     if(version VERSION_LESS PYTHON_MINIMUM_REQUIRED)
-       set(PYTHON_PREFER_EXECUTABLE "PYTHON_PREFER_EXECUTABLE-NOTFOUND")
-     else()
-       set(PYTHON_MINIMUM_REQUIRED ${version})
-       set(PYTHON_EXACT EXACT)
-       # Python3_ROOT_DIR ensures that location will be preferred by FindPython3.
-       # On Linux, this has no impact as it will usually be /usr/bin
-       # but on Windows it solve issues when both 32 and 64 bit versions are
-       # installed, as version is not enough and FindPython3 might pick the
-       # version not on %PATH%. Setting Python3_ROOT_DIR ensures we are using
-       # the version we just tested.
-       get_filename_component(PYTHON_PATH ${PYTHON_PREFER_EXECUTABLE} DIRECTORY)
-       set(Python3_ROOT_DIR ${PYTHON_PATH})
-       break()
-     endif()
+# If we're told exactly which Python executable to use, we should respect that,
+# instead of using the default matching process below.
+if(DEFINED PYTHON_EXECUTABLE)
```

Comment:
back in the rabbit hole.

Zephyr has been struggling with inconsistent CMake behavior on the Python lookup as well as a request to use the same Python interpreter as `west` whenever `west build` is invoked.

Part of this work resulted in `PYTHON_PREFER`, both to ensure that Zephyr would keep looking for a better alternative, but also because `PYTHON_EXECUTABLE` as input was not supported until CMake >= 3.16:
> New in version 3.16.
> To solve special cases, it is possible to specify directly the artifacts by setting the following variables:
>
> `Python_EXECUTABLE`
>     The path to the interpreter.

Ref: https://cmake.org/cmake/help/latest/module/FindPython.html#artifacts-specification

Now minimum CMake required by Zephyr is 3.20, I would therefore like to see this code be cleaned up instead of more patching.

Note. if you set `PYTHON_PREFER` to absolute path, such as `PYTHON_PREFER=/use/this/python`, then it should behave identical to `PYTHON_EXECUTABLE` (which couldn't be used as input back then).
Only difference would be the `FATAL_ERROR`, and that could be done without all the additional changes.

Thus, a request to see if `PYTHON_PREFER` can be deprecated in favor of `PYTHON_EXECUTABLE`, and thereby simplify the code.

Please take a look at the history, some PRs and issues:
PRs:
- https://github.com/zephyrproject-rtos/zephyr/pull/24308
- https://github.com/zephyrproject-rtos/zephyr/pull/34479

Issues:
- https://github.com/zephyrproject-rtos/zephyr/issues/24340
- https://github.com/zephyrproject-rtos/zephyr/issues/23922
- https://github.com/zephyrproject-rtos/zephyr/issues/24252
- https://github.com/zephyrproject-rtos/zephyr/issues/11103
-